Level 3

AndreiZhura opened the pull request.

AndreiZhura committed:
level-3 first commit

AndreiZhura committed:
fix start

AndreiZhura committed:
protectedRouter

AndreiZhura committed:
protectedRouter

AndreiZhura committed:
fix:ProtectedRoute

AndreiZhura committed:
movies

AndreiZhura committed:
fix:Registration

AndreiZhura committed:
fix: register and Login

AndreiZhura committed:
1

AndreiZhura committed:
navigate

AndreiZhura committed:
1

AndreiZhura committed:
fix Register

AndreiZhura committed:
завтра продолжу

AndreiZhura committed:
fix: что-то пошло не так

AndreiZhura committed:
1

AndreiZhura committed:
fix: level-3

AndreiZhura committed:
fix: CurrentUser

AndreiZhura committed:
fix: CurrentUser

AndreiZhura committed:
token

AndreiZhura committed:
работает редактирование профиля

AndreiZhura committed:
movies

AndreiZhura committed:
Movies

AndreiZhura committed:
fix: 1

AndreiZhura committed:
завтра продолжу

AndreiZhura committed:
получилось добавление карточек по нажатию на кнопку

AndreiZhura committed:
кнопка исчезает когда все фильмы выгружены

AndreiZhura committed:
настроил хедер подправил dthcnre

AndreiZhura committed:
Preloader

AndreiZhura committed:
fix: checkbox

AndreiZhura committed:
fix: checkbox

AndreiZhura committed:
заработали короткометражки

AndreiZhura committed:
fix: check

AndreiZhura committed:
1

AndreiZhura committed:
1

AndreiZhura committed:
завтра продолжу

AndreiZhura committed:
сделал

AndreiZhura committed:
переделываю поиск

AndreiZhura committed:
filterMovies

AndreiZhura committed:
поиск заработал корректно

AndreiZhura committed:
начинаю делать валидацию поиска

AndreiZhura committed:
fix: сделал валидацию формы

AndreiZhura committed:
fix: сделал прелоадер

AndreiZhura committed:
сделал ошибку валидации

AndreiZhura committed:
fix: ничего не найдено

AndreiZhura committed:
fix: теперь на странице показывается что на сервере произошла ошибка

AndreiZhura committed:
реализовал переключение поиска с коротких на длинные

AndreiZhura committed:
setWidth

AndreiZhura committed:
resize начал делать

AndreiZhura committed:
настроил количество карточек под разрешение

AndreiZhura committed:
начал делать сохранение карточек

AndreiZhura committed:
делаю лайки

AndreiZhura committed:
работает добавление фильмов

AndreiZhura committed:
1

AndreiZhura committed:
решил сделать заново савес кард

AndreiZhura committed:
1

AndreiZhura committed:
1

AndreiZhura committed:
проверяю что получилось

AndreiZhura committed:
отправная точка

AndreiZhura committed:
fix: реализовал удаление карточки

AndreiZhura committed:
решил попробовать savedMovies отдельно

AndreiZhura committed:
переделал saves

AndreiZhura committed:
fix: работает удаление и добавление карточек

AndreiZhura committed:
отправная точка

AndreiZhura committed:
fix: реализовал удаление карточки в режиме реального времени

AndreiZhura committed:
отправная точка

AndreiZhura committed:
fix: получилось окрасить кнопку лайка

AndreiZhura committed:
начал реализовывать кнопку лайка

AndreiZhura committed:
сделал удаление карточек

AndreiZhura committed:
поиск и короткометражки работает в сохраненках

AndreiZhura committed:
ничего не найдено

AndreiZhura committed:
localstorage

AndreiZhura committed:
localstorage

AndreiZhura committed:
fix: подправил кнопку поиска

AndreiZhura committed:
пофиксил localStorage

AndreiZhura committed:
пофиксил баг

AndreiZhura committed:
начал делать валидацию отправная точка

AndreiZhura committed:
работает валидация логина

AndreiZhura committed:
валидация работает login and register

AndreiZhura committed:
пофиксил баг userInform

AndreiZhura committed:
проверка перед ревью

AndreiZhura committed:
данный пользователь уже сушествует

AndreiZhura committed:
fix: выводится ошибка логина

AndreiZhura committed:
валидация профиля

AndreiZhura committed:
пофиксил баг с карточками

AndreiZhura committed:
перед первым ревью

AndreiZhura committed:
README.md

AndreiZhura committed:
README.md

AndreiZhura committed:
перед ревью

AndreiZhura committed:
перед ревью

AndreiZhura committed:
перед ревью

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.
    **Комментарий: пользователь не видит точного результата своего последнего поиска на странице фильмов при повторном переходе на эту страницу с другой страницы и после перезагрузки сайта - исчезает строка поиска в форме поиска.**
    
    **Напоминание: для того, чтобы выполнить вышележащий пункт чек-листа, при каждом поиске на странице фильмов результаты поиска фильмов нужно сохранять в localstorage. В сохраняемые данные должны входить и строка поиска и статус чекбокса «Короткометражки», чтобы пользователю при повторном переходе на страницу фильмов, можно было вывести на экран информацию, полностью совпадающую с его последним поиском  перед уходом на другую страницу или перезагрузкой страницы (с расстановкой актуального состояния лайков). То есть, сохранённую информацию можно будет доставать из хранилища при монтировании и перерисовке компоненты фильмов и выводить на экран. При логауте пользователя нужно эту информацию в хранилище  и стейт-переменных из useState уничтожать.**    
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.   
	**Комментарий: Если запрос прошёл успешно, то происходит редирект на страницу авторизации, а нужен автоматический вход на `/movies` как авторизованного пользователя, без необходимости посещения страницы авторизации.**
	
	**Комментарий: На странице «Регистрация» при попытке пользователя зарегистрировать адрес почты, который уже зарегистрирован, сервер приложения присылает в ответе ошибку 500. Но, если почта уже была зарегистрирована, сервер приложения должен присылать ошибку 409 в ответе.**
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
    **Комментарий: чтобы авторизоваться на странице «Авторизация» нужно послать запрос к серверу приложения два раза. После первого запроса сервер приложения присылает ошибки 401 при обращении к ендпоинтам me и movies (см. снимок), и пользователь так и остаётся на странице «Авторизация». После отсылки второго запроса пользователем вход происходит. Кейс необходимо протестировать, можно на пользователях review1@gmail.com, review2@gmail.com, пароль 12345678.**
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
  **Комментарий: Нужно, чтобы в личном кабинете отображались текущие данные пользователя, а не плейсхолдеры, чтобы пользователь мог данные редактировать, а не вводить снова.**
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    **Комментарий: На странице профиля при попытке пользователя сохранить адрес почты, который принадлежит другому пользователю, сервер приложения присылает в ответе ошибку 500. Но, если почта уже была зарегистрирована другим пользователем, сервер приложения должен присылать ошибку 409 в ответе, пользователю нужно выдать об этом сообщение.**
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
    **Комментарий: Пользователю не отображается уведомление об успешном запросе к серверу при сохранении профиля.**
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.
    **Комментарий: пользователь может сохранить текущие данные профиля, не изменяя их, а нужно сделать так, чтобы можно было отправить запрос на сохранение, только если  данные изменены по сравнению с текущими данными пользователя, которые должны содержаться в текущий момент в контексте.**
    
    **Напоминание: после каждого успешного сохранения данных профиля на сервере, нужно эти данные вносить в контекст юзера.**
    
    **Замечание: Чтобы данные профиля считались изменёнными, пользователю должно быть достаточно изменения хотя бы одного данного. Может быть изменён один любой инпут из двух, и отправлен запрос на сервер.**  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
	**Комментарий: Надпись "Ничего не найдено" присутствует на странице фильмов сразу же при входе, когда пользователь ещё ничено не искал, а не только когда ничего не найдено.**
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [ ] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  **Комментарий: При попытке авторизованного пользователя войти на любую страницу по её URL-адресу в адресной строке, происходит вход пользователя на страницу «Сохранённые фильмы», а не на эти страницы, хотя редирект не по набранноу адресу должен происходить только при попытке авторизованного пользователя выйти на страницы авторизации и регистрации.**
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  **Комментарий: После успешного вызова обработчика `onSignOut` происходит редирект на страницу авторизации, а не на `/`.**
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  **Комментарий: авторизованный пользователь не может выйти на страницу «404».**
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [x] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 

### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 53.
**P.S.**
**Все исправления кода должны быть видны в коммитах в репозитории. В том числе должен быть виден и коммит по исправлению бэкенда, когда сервер приложения присылает в ответе ошибку 500 вместо ошибки 409 (см. комментарии в Резюме по этому поводу). `Укажите ссылку на репозиторий с бэкендом`.** 
**На исправление каждой ошибки должен быть отдельный коммит. Сделайте, пожалуйста, у каждого коммита понятное название, говорящее, что коммит исправляет. Это обязательные требования.**

AndreiZhura committed:
в readme указал ссылку на репозиторий с бэкендом

AndreiZhura committed:
BPVTYBK HTLBHTRN YF CNHFYBWE FDNJHBPFWBB

AndreiZhura committed:
Комментарий: пользователь не видит точного результата своего последнего поиска на странице фильмов при повторном переходе на эту страницу с другой страницы и после перезагрузки сайта - исчезает строка поиска в форме поиска. (исправил)

AndreiZhura commented:
Уважаемая Ирина Мозес
Доброго времени суток !

я проверил Ваше замечание : Комментарий: На странице «Регистрация» при попытке пользователя зарегистрировать адрес почты, который уже зарегистрирован, сервер приложения присылает в ответе ошибку 500. Но, если почта уже была зарегистрирована, сервер приложения должен присылать ошибку 409 в ответе.

![image](https://user-images.githubusercontent.com/95981207/220167859-d8564912-3cd3-4cbc-9436-868651f5acf7.png)

AndreiZhura committed:
Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-фай && Комментарий: Если запрос прошёл успешно, то происходит редирект на страницу авторизации, а нужен автоматический вход на /movies как авторизованного пользователя, без необходимости посещения страницы авторизации.

AndreiZhura committed:
убрал константы

AndreiZhura committed:
Комментарий: Нужно, чтобы в личном кабинете отображались текущие данные пользователя, а не плейсхолдеры, чтобы пользователь мог данные редактировать, а не вводить снова.

AndreiZhura committed:
(исправил)  Комментарий: Нужно, чтобы в личном кабинете отображались текущие данные пользователя, а не плейсхолдеры, чтобы пользователь мог данные редактировать, а не вводить снова.

На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут /users/me, если данные введены корректно.
Комментарий: На странице профиля при попытке пользователя сохранить адрес почты, который принадлежит другому пользователю, сервер приложения присылает в ответе ошибку 500. Но, если почта уже была зарегистрирована другим пользователем, сервер приложения должен присылать ошибку 409 в ответе, пользователю нужно выдать об этом сообщение.
Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
Комментарий: Пользователю не отображается уведомление об успешном запросе к серверу при сохранении профиля.
Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.
Комментарий: пользователь может сохранить текущие данные профиля, не изменяя их, а нужно сделать так, чтобы можно было отправить запрос на сохранение, только если данные изменены по сравнению с текущими данными пользователя, которые должны содержаться в текущий момент в контексте.

AndreiZhura committed:
Если ничего не найдено выводится надпись ничего не найдено а в самом начале Начните поиск

AndreiZhura committed:
исправил (Комментарий: После успешного вызова обработчика onSignOut происходит редирект на страницу авторизации, а не на /.)

AndreiZhura committed:
начните поиск || ничего не найдено исправил баг

AndreiZhura committed:
(исправленно) При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.

AndreiZhura committed:
После успешного вызова обработчика onSignOut происходит редирект на /.
Комментарий: После успешного вызова обработчика onSignOut происходит редирект на страницу авторизации, а не на /. (исправленно)

AndreiZhura committed:
1

AndreiZhura committed:
1

AndreiZhura committed:
Комментарий: Если запрос прошёл успешно, то происходит редирект на страницу авторизации, а нужен автоматический вход на /movies как авторизованного пользователя, без необходимости посещения страницы авторизации. (исправленно)

AndreiZhura committed:
немного отредактировал регистрацию

AndreiZhura committed:
немного отредактировал регистрацию

AndreiZhura committed:
немного отредактировал регистрацию

AndreiZhura committed:
немного отредактировал регистрацию

AndreiZhura committed:
поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего. (при отправке запроса кнопка не кликабельна)

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.
    **Комментарий: Сохранять данные последнего поиска на странице «Сохранённые фильмы» в localstorage не нужно. Сейчас строка поиска, заданный на странице «Сохранённые фильмы», сохраняется в localstorage в той же переменной search, что и строка поиска со страницы фильмов, заменяя её значение, что ведёт к неправильному повторному показу результата поиска на странице фильмов. Нужно, чтобы при входе на страницу «Сохранённые фильмы» пользователь видел все свои сохранённые карточки без задания параметров поиска. Функциолнал необходимо протестировать, можно на пользователе review1@gmail.com, пароль 12345678, у которого есть сохранённые карточки со слогом "ро" в названии.**
    
    **Комментарий: Функционалы поиска на страницах «Фильмы» и «Сохранённые фильмы» никак не должны быть связаны, они должны быть полностью не зависимы друг от друга.**
     
    **Комментарий: сохранение статуса чекбокса при поиске в переменной shortORlong из localstorage происходит непредсказуемо, что также ведёт к неправильному повторному показу результата поиска на странице фильмов. Статус чекбокса в этой переменной должен сохраняться при каждом изменении результата поиска на экране страницы фильмов. При этом, если чекбокс устанавливается пользователем в положение "короткометражки", в shortORlong должно сохраняться true. Если чекбокс устанавливается в другое положение, то в shortORlong должно сохраняться false. Дефолтным статусом чекбокса должно быть значение false. Функциолнал необходимо протестировать, можно на пользователе review1@gmail.com, пароль 12345678, у которого есть сохранённые карточки со слогом "ро" в названии.** 
    
    **Напоминание: для того, чтобы выполнить вышележащий пункт чек-листа, при каждом поиске на странице фильмов результаты поиска фильмов нужно сохранять в localstorage. В сохраняемые данные должны входить и строка поиска и статус чекбокса «Короткометражки», чтобы пользователю при повторном переходе на страницу фильмов, можно было вывести на экран информацию, полностью совпадающую с его последним поиском  перед уходом на другую страницу или перезагрузкой страницы (с расстановкой актуального состояния лайков). То есть, сохранённую информацию можно будет доставать из хранилища при монтировании и перерисовке компоненты фильмов и выводить на экран. При логауте пользователя нужно эту информацию в хранилище  и стейт-переменных из useState уничтожать.**    
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.   	
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.
   **Комментарий: При редактировании пользователь может ввести текущие данные профиля и сохранить их, а нужно сделать так, чтобы можно было отправить запрос на сохранение, только если  данные изменены по сравнению с текущими данными пользователя, которые должны содержаться в текущий момент в контексте.**
    
    **Напоминание: после каждого успешного сохранения данных профиля на сервере, нужно эти данные вносить в контекст юзера.**
    
    **Замечание: Чтобы данные профиля считались изменёнными, пользователю должно быть достаточно изменения хотя бы одного данного. Может быть изменён один любой инпут из двух, и отправлен запрос на сервер.**  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  **Комментарий: При попытке авторизованного пользователя войти на любую страницу по её URL-адресу в адресной строке, происходит вход пользователя на главную страницу, а не на эти страницы, хотя редирект не по набранноу адресу должен происходить только при попытке авторизованного пользователя выйти на страницы авторизации и регистрации.**
  
  **Комментарий: По URL-адресу авторизованный пользователь может зайти и на страницы авторизации и регистрации, что не допустимо, так как открывается возможность авторизоваться ещё раз, не выйдя из предыдущего аккаунта.**
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [x] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
  **Комментарий: При запросах на авторизацию и регистрацию, разблокировать формы нужно только в блоке catch, а не в finally. Иначе, после завершения успешного запроса, можно ещё раз успеть послать новый запрос перед переходом на страницу фильмов.**
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: длительность короткометражек - число 40, нужно занести в константу с понятным названием и поместить в файл констант, чтобы в коде не было "магических" значений. Смысл литеральных числовых значений  (таких как 40, или любых других чисел), встречающихся в коде, не возможно понять, если их не занести в константы с "говорящим названием" и использовать вместо них эти константы в коде. Поэтому числа и называются "магическими" значениями.**
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.
## Доп. Комментарии:
**Комментарий: : Ели у пользователя есть сохранённые фильмы, при входе на страницу «Сохранённые фильмы» надо выводить на экран все сохранённые карточки, так как пользователь может не помнить, какие фильмы он сохранял, и не найти их.  Нужно, чтобы при входе на эту страницу пользователь видел все свои избранные фильмы без задания параметров поиска, а  не пустой экран и не результат какого-либо поиска .**
## Количество баллов: 53.
6 баллов снято за замечание из комментария в разделе "Доп. Комментарии".
**P.S.**
**Все исправления кода должны быть видны в коммитах в репозитории. На исправление каждой ошибки должен быть отдельный коммит. Сделайте, пожалуйста, у каждого коммита понятное название, говорящее, что коммит исправляет. Это обязательные требования.**

AndreiZhura committed:
Комментарий: Функционалы поиска на страницах «Фильмы» и «Сохранённые фильмы» никак не должны быть связаны, они должны быть полностью не зависимы друг от друга . (сделанно)

AndreiZhura committed:
отменяю прошлый коммит

AndreiZhura committed:
Комментарий: Функционалы поиска на страницах «Фильмы» и «Сохранённые фильмы» никак не должны быть связаны, они должны быть полностью не зависимы друг от друга.(теперь поиск полностью независим так же у сохраненных фильмах не сохраняется чекбок короткметражек)

AndreiZhura committed:
Комментарий: сохранение статуса чекбокса при поиске в переменной shortORlong из localstorage происходит непредсказуемо (исправленно теперь в localstorage корректно сохраняются а затем выводятся данные)

AndreiZhura committed:
Комментарий: При редактировании пользователь может ввести текущие данные профиля и сохранить их, а нужно сделать так, чтобы можно было отправить запрос на сохранение, только если данные изменены по сравнению с текущими данными пользователя, которые должны содержаться в текущий момент в контексте . (исправленно)

AndreiZhura committed:
Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл . теперь отдельные числовые значения обозначенны константами и перенесенны в отдельный файл

AndreiZhura committed:
1.По URL-адресу авторизованный пользователь может зайти и на страницы авторизации и регистрации, что не допустимо (исправленно) 2.При попытке авторизованного пользователя войти на любую страницу по её URL-адресу в адресной строке, происходит вход пользователя на главную страницу, а не на эти страницы, хотя редирект не по набранноу адресу должен происходить только при попытке авторизованного пользователя выйти на страницы авторизации и регистрации . (исправленно)

AndreiZhura committed:
Комментарий: При запросах на авторизацию и регистрацию, разблокировать формы нужно только в блоке catch, а не в finally. Иначе, после завершения успешного запроса, можно ещё раз успеть послать новый запрос перед переходом на страницу фильмов. (блок finnally удален)

AndreiZhura committed:
прохожусь по чек листу: длитильность фильма соответствует макету

AndreiZhura committed:
корректировка

AndreiZhura committed:
Вернул регулярку на место

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
	**Комментарий: при входе на страницы «Регистрация» и «Авторизация» можно послать запрос на сервер с пустыми полями ввода (см. снимок), что противоречит вышележащему пункту чек-листа.**
![2023-02-27_13-30](https://user-images.githubusercontent.com/19558891/221513822-88c49dd7-1afb-45b5-ae6d-8a3b6f78f09e.png)
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.   	
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.
   **Комментарий: кнопка "Редактировать" не всегда правильно меняет свой статус активности (становится доступной или недоступной для сабмита). В каких-то случаях (закономерность не просматривалась) удавалось сохранить неизменённые данные профиля.**  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
  **Комментарий: На время выполнения запроса считается хорошей практикой блокировать кнопку отправки формы (поля ввода можно не блокировать), чтобы пользователь не мог нажать эту кнопку ещё раз во время  выполнения запроса, до его завершения. Блокировать кнопку отправки формы нужно перед отправкой запроса на сервер, то есть в первой командной инструкции в методах-обработчиках сабмита формы. В блоке catch этих методов кнопку сабмита нужно разблокировать, чтобы, в случае отрицательного ответа сервера, пользователь мог попытаться отправить данные на сервер ещё раз. Сейчас можно успеть ещё раз послать новый запрос во время выполнения уже посланного запроса на авторизацию или регистрацию.**
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.
## Количество баллов: 91.
7 баллов снято за замечание из комментария в разделе "Функциональность проекта из брифа реализована полностью".

## У Вас очень хороший дипломный проект!

Чтобы красиво оформить файл README.md, где рассказать о собственной работе над проектом и собственных достижениях, можно воспользоваться инструментом Markdown. Описание языка Markdown есть в Ваших лекциях по этой ссылке:
https://praktikum.yandex.ru/learn/web/courses/dbf98e55-0f76-444b-850c-4538708ad571/sprints/1425/topics/b4072eed-2089-45c0-9382-98ea71202341/lessons/939aeab7-0508-49e8-bcf2-199d4dbf74f7/.

Ссылки на интересные материалы и подкасты:
[habr](https://habr.com/ru/)
https://soundcloud.com/everyonecan
https://soundcloud.com/podlodka
https://soundcloud.com/devschacht
https://soundcloud.com/begebot
https://soundcloud.com/web-standards
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw

Интересная статья о замыкании и лексическом окружении функций:
https://developer.mozilla.org/ru/docs/Web/JavaScript/Closures.

О пользовательских хуках:
https://ru.reactjs.org/docs/hooks-custom.html

https://reactdev.ru/
Справочник React

Какие бывают стили в названиях файлов js и зачем они нужны, интересно и кратко написано здесь:
https://medium.com/@alivander/camel-pascal-snake-case-и-другие-стили-написания-288ec62ca0d0

Ресурс по библиотеке Redux, которая очень широко используется в реальных проектах на React:
https://rajdee.gitbooks.io/redux-in-russian/content/

Создание приложений JavaScript с помощью TypeScript
https://docs.microsoft.com/ru-ru/learn/paths/build-javascript-applications-typescript/

Руководство Typescript
http://typescript-lang.ru/docs/

TypeScript в деталях.
https://typescript-definitive-guide.ru/book/chapters

**Желаю дальнейших творческих успехов, экспертных знаний и удачи!**
